### PR TITLE
xworld3d with simple navigation

### PR DESCRIPTION
**Pull request by skylian**
xworld 3d environment based on roboschool

### CODE REVIEW

File: games/xworld3d/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,82 @@
+function(append_libraries lib_list library_dir libraries)
```
Comment (yu239) [inline games/xworld3d/CMakeLists.txt]: Put Bullet Physics, Assimp, and Qt5 into separate cmake files and include them.

File: games/xworld3d/x3item.h [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+
+#define EPSILON 1e-6
+#define UNIT 0.1
+
+using roboschool::Pose;
+using roboschool::Object;
+using roboschool::World;
+using roboschool::Camera;
+
+struct Loc3 {
```
Comment (yu239) [inline games/xworld3d/x3item.h]: Consider using Vec3 defined simulator_entity.h

File: games/xworld3d/x3item.h [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+               const std::vector<int>& v) :
+            name(n), type(t), model_file(f), loc(v) {}
+
+    std::string name;
+    X3EntityType type;
+    std::string model_file;
+    Loc3 loc;
+};
+
+class X3Item : public X3Entity {
```
Comment (yu239) [inline games/xworld3d/x3item.h]: Unify all Items.

File: games/xworld3d/x3item.h [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+    X3Agent* agent_;
+};
+
+}} // simulator::xworld3d
+
+namespace std {
+template <>
+struct hash<simulator::xworld3d::Loc3> {
+    size_t operator()(const simulator::xworld3d::Loc3& l) const {
+        return hash<int>()(l.x) ^ hash<int>()(l.y);
```
Comment (yu239) [inline games/xworld3d/x3item.h]: Where is l.z?

File: games/xworld3d/x3item.h [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+
+typedef std::shared_ptr<X3Agent> X3AgentPtr;
+
+/********************************** X3Camera **********************************/
+class X3Camera {
+public:
+    X3Camera(World& world, int img_height, int img_width);
+
+    Pose pose() { return camera_.pose(); }
+
```
Comment (yu239) [inline games/xworld3d/x3item.h]: Add some comments for this class.

File: games/xworld3d/x3item.h [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+    void detach() { agent_ = NULL; }
+
+private:
+    void attach_agent(X3Agent* agent);
+
+    void update(bool is_debug);
+
+    Camera camera_;
+    X3Agent* agent_;
```
Comment (yu239) [inline games/xworld3d/x3item.h]: Avoid using raw pointers.

File: games/xworld3d/x3scene.cpp [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+        add_item(i);
+    }
+    CHECK_GT(agents_.size(), 0) << "There should be at least one agent.";
+}
+
+void X3Scene::add_item(const X3ItemInfo& info) {
+    if (info.type == AGENT) {
+        CHECK(agents_.find(info.name) == agents_.end())
+                << "An agent with name " << info.name << " exists.";
+        auto agent = std::make_shared<X3Agent>(info, world_,
```
Comment (yu239) [inline games/xworld3d/x3scene.cpp]: Potentially other items could have speed

File: games/xworld3d/x3scene.cpp [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+            break;
+        case X3NavAction::TURN_RIGHT:
+            agent->turn_right();
+            break;
+        case X3NavAction::JUMP:
+            agent->jump();
+            break;
+        case X3NavAction::COLLECT:
+            {
+                X3Goal* goal = collect_goal(agent);
```
Comment (yu239) [inline games/xworld3d/x3scene.cpp]: avoid using raw pointers

File: games/xworld3d/x3scene.cpp [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+}
+
+roboschool::RenderResult X3Scene::render(std::string agent_name, bool debug) {
+    auto it = agents_.find(agent_name);
+    CHECK(it != agents_.end())
+            << "agent " << agent_name << " not found.";
+    return camera_->render(it->second.get(), debug);
+}
+
+void X3Scene::remove_goal(X3Goal* goal) {
```
Comment (yu239) [inline games/xworld3d/x3scene.cpp]: Do we also have remove_agent, remove_block?

File: games/xworld3d/x3scene.cpp [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+        auto& g = it->second;
+        world_.remove_object(g->object());
+        g->destroy();
+        goals_.erase(it);
+    }
+}
+
+X3Goal* X3Scene::collect_goal(const X3AgentPtr& agent) {
+    X3Goal* goal = NULL;
+    float best_score = 0.5;
```
Comment (yu239) [inline games/xworld3d/x3scene.cpp]: Explain why the best score is 0.5

File: games/xworld3d/x3scene.h [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+    int img_height() const { return img_height_; }
+    int img_width() const { return img_width_; }
+
+    void build_scene(const std::vector<X3ItemInfo>& infos,
+                     const std::string& floor_file,
+                     const std::string& stadium_file);
+
+    void add_item(const X3ItemInfo& info);
+
+    void act(std::string agent_name, int a, float& reward, int& event) {
```
Comment (yu239) [inline games/xworld3d/x3scene.h]: Why not use agent_id instead of agent_name

File: games/xworld3d/xworld3d_simulator.cpp [hunk 12]
```diff
@@ -0,0 +1,382 @@
...
+    ~X3SimulatorImpl() {}
+
+    std::string name() { return agent_name_; }
+
+    void reset_game();
+
+    // game over when either reaching goal or passing the number of max steps
+    int game_over();
+
+    int max_steps() {
```
Comment (yu239) [inline games/xworld3d/xworld3d_simulator.cpp]: This should be more general if the world is continuous.

File: games/xworld3d/xworld3d_simulator.cpp [hunk 13]
```diff
@@ -0,0 +1,382 @@
...
+    img_width_ = x3scene_.img_width();
+    img_height_out_ = world_height_ * GRID_SIZE;
+    img_width_out_ = world_width_ * GRID_SIZE;
+    curr_screen_.resize(3 * img_width_ * img_height_, 0);
+
+    agent_name_ = x3parser_.assign_agent();
+    event_of_last_action_ = X3Event::NOTHING;
+}
+
+int X3SimulatorImpl::game_over() {
```
Comment (yu239) [inline games/xworld3d/xworld3d_simulator.cpp]: Maybe game_over should be defined by python task. Shouldn't be hard coded.

File: games/xworld3d/xworld3d_simulator.h [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+
+#include "simulator.h"
+#include "xworld3d_flags.h"
+
+namespace simulator {
+namespace xworld3d {
+
+class X3SimulatorImpl;
+
+class X3Simulator : public GameSimulator {
```
Comment (yu239) [inline games/xworld3d/xworld3d_simulator.h]: Why not inherit from MultiGameSimulator?

File: games/xworld3d/x3item.h (hunk 6 above)
Comment (skylian) [reply games/xworld3d/x3item.h]: using raw pointers is ok here as the class X3Camera does not own X3Agent

File: games/xworld3d/x3item.h (hunk 4 above)
Comment (skylian) [reply games/xworld3d/x3item.h]: now we use Vec3, we need a better hash function

File: games/xworld3d/x3scene.cpp (hunk 8 above)
Comment (skylian) [reply games/xworld3d/x3scene.cpp]: we should use raw pointers here as we dont want to claim ownership here

File: games/xworld3d/xworld3d_simulator.h (hunk 14 above)
Comment (skylian) [reply games/xworld3d/xworld3d_simulator.h]: i thought we were going to abandon MultiGameSimulator

File: games/xworld3d/x3scene.cpp (hunk 10 above)
Comment (skylian) [reply games/xworld3d/x3scene.cpp]: if no score is higher than 0.5, we consider the collecting action failed

File: games/xworld3d/x3scene.h (hunk 11 above)
Comment (skylian) [reply games/xworld3d/x3scene.h]: if we use agent_id, we still need a map from id to agent name

File: games/xworld3d/x3scene.cpp (hunk 9 above)
Comment (skylian) [reply games/xworld3d/x3scene.cpp]: not now. adding them is easy.

**Comment by skylian** [comment]
for questions left unanswered, i'll take them into next PR

File: games/xworld3d/xworld3d_flags.cpp [hunk 15]
```diff
@@ -0,0 +1,25 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "xworld3d_flags.h"
+
+DEFINE_string(x3_glsl_path, "", "directory path to OpenGL shader scripts");
+DEFINE_string(x3_model_dir, "", "3D model directory");
+DEFINE_string(x3_conf, "", "Xworld 3D configuration file");
```
Comment (Haichao-Zhang) [inline games/xworld3d/xworld3d_flags.cpp]: Xworld -> XWorld

File: games/xworld3d/xworld3d_flags.cpp [hunk 16]
```diff
@@ -0,0 +1,25 @@
...
+DEFINE_int32(x3_img_width, 256, "width of the rendering image");
+DEFINE_int32(x3_img_height, 256, "height of the rendering image");
+DEFINE_double(x3_reaching_distance, 0.12, "");
```
Comment (Haichao-Zhang) [inline games/xworld3d/xworld3d_flags.cpp]: Add explanation to x3_reaching_distance in "" ?

File: games/xworld3d/xworld3d_parser.cpp [hunk 17]
```diff
@@ -0,0 +1,218 @@
...
+    // TODO: use a class to pre-load these
+    item_repo_["ground"] = path_join({model_root_dir_, "floor.xml"});
+    item_repo_["stadium"] = path_join({model_root_dir_, "stadium/stadium.obj"});
+    item_repo_["block"] = path_join({model_root_dir_, "block/block.urdf"});
+    item_repo_["agent"] = path_join({model_root_dir_, "ball/ball.urdf"});
+    item_repo_["goal"] = path_join({model_root_dir_, "apple/apple.urdf"});
+}
+
+void X3Parser::reset_config(const std::string& world_config, bool print) {
+    curriculum_.num_games_so_far ++;
```
Comment (Haichao-Zhang) [inline games/xworld3d/xworld3d_parser.cpp]: no need to add space before ++

File: games/xworld3d/xworld3d_simulator.cpp [hunk 18]
```diff
@@ -0,0 +1,382 @@
...
+    cv::resize(img, img1, img1.size(), cv::INTER_LINEAR);
+    get_screen_rgb(img, true);
+    cv::resize(img, img2, img2.size(), cv::INTER_LINEAR);
+
+    return concat_images(img1, img2, false);
+
+    // cv::Mat img;
+    // get_screen_rgb(img);
+    //
+    // return img;
```
Comment (Haichao-Zhang) [inline games/xworld3d/xworld3d_simulator.cpp]: line 287-290 can be removed

File: games/xworld3d/x3item.cpp [hunk 19]
```diff
@@ -0,0 +1,144 @@
...
+        camera_(world.new_camera_free_float(img_width, img_height, "camera")),
+        agent_(NULL) {}
+
+void X3Camera::attach_agent(X3Agent* agent) {
+    if (agent && agent_ != agent) {
+        agent_ = agent;
+    }
+}
+
+void X3Camera::update(bool debug) {
```
Comment (Haichao-Zhang) [inline games/xworld3d/x3item.cpp]: suggest to change variable debug to a better name reflecting it corresponds to bird_view if turned on.

File: games/xworld3d/x3item.cpp [hunk 20]
```diff
@@ -0,0 +1,144 @@
...
+        agent_->get_direction(dir_x, dir_y);
+        camera_.move_and_look_at(p.x(), p.y(), p.z(),
+                                 p.x() + dir_x, p.y() + dir_y, p.z());
+    } else {
+        // bird view
+        camera_.move_and_look_at(p.x(), p.y(), 1.0f, p.x(), p.y(), p.z());
+    }
+}
+
+roboschool::RenderResult X3Camera::render(X3Agent* agent, bool is_debug) {
```
Comment (Haichao-Zhang) [inline games/xworld3d/x3item.cpp]: suggest to change variable "is_debug" to a better name reflecting it corresponds to bird_view if turned on.

File: games/xworld3d/x3item.h [hunk 21]
```diff
@@ -0,0 +1,265 @@
...
+    Pose pose() { return camera_.pose(); }
+
+    roboschool::RenderResult render(X3Agent* agent, bool debug = false);
+
+    void detach() { agent_ = NULL; }
+
+private:
+    void attach_agent(X3Agent* agent);
+
+    void update(bool is_debug);
```
Comment (Haichao-Zhang) [inline games/xworld3d/x3item.h]: is_debug and debug in line 243, either change to better names or add explanation on their functionality.

File: games/xworld3d/x3item.h [hunk 22]
```diff
@@ -0,0 +1,197 @@
...
+    X3Agent* agent_;
+};
+
+}} // simulator::xworld3d
+
+// TODO: need a better hash function
+namespace std {
+template <>
+struct hash<simulator::Vec3> {
+    size_t operator()(const simulator::Vec3& l) const {
```
Comment (yu239) [inline games/xworld3d/x3item.h]: make sure you fix this before checking in. This error will be a disaster if left for the future.

**Pull request merged**